File: pythonStudy/BOJ/1708 볼록 껍질.py
# 1708 볼록 껍질
from functools import cmp_to_key
input = __import__('sys').stdin.readline

MXN = 40_001
def ccw(x1, y1, x2, y2, x3, y3):
    return (x2 - x1) * (y3 - y1) - (y2 - y1) * (x3 - x1)
def subVec(a, b):
    return (a[0] - b[0], a[1] - b[1])
def sortKey(itm):
    global cp
    x, y = subVec(itm, cp)
    if x:
        return (y / x, x, y)  # floating point
    else:
        return (2 * MXN, x, y)
def cmp(itm1, itm2):
    cc = ccw(*cp, *itm1, *itm2)
    if (cc): return -cc
    # return sum(itm1) - sum(itm2)
    if (itm1[0] - itm2[0]): return itm1[0]-itm2[0]
    return itm1[1]-itm2[1]

pts = []
cp = (MXN, MXN)  # control point
ptn = 0
for _ in range(int(input())):
    a, b = map(int, input().split())
    if (a, b) < cp:
        if cp[0] != MXN: pts.append(cp)
        cp = (a, b)
    else:
        pts.append((a, b))

pts.sort(key=cmp_to_key(cmp)) # or key = sortKey
print(cp, pts)
stk = [cp, pts[0]]
for idx in range(1, len(pts)):
    pt = pts[idx]
    while (2 <= len(stk) and ccw(*stk[-2], *stk[-1], *pt) <= 0):
        stk.pop()
    stk.append(pt)
# print(stk)
print(len(stk))

File: C++/Solved/BOJ/1708 볼록 껍질.cpp
#include <bits/stdc++.h>
using namespace std;
// 1708 볼록 껍질
#define FOR(i, n, m) for(int i=(n); i<(m); i++)
#define X first
#define Y second
using ll = long long;
using PII = pair<int, int>;

int ptN;
PII pts[100002];
PII stk[100002];
int stkIdx;
ll ccw(PII p1, PII p2, PII p3) {
	return 1ll * (p2.X - p1.X) * (p3.Y - p1.Y) -
		1ll * (p2.Y - p1.Y) * (p3.X - p1.X);
}
bool cmp(PII v1, PII v2) {
	ll cc = ccw(pts[0], v1, v2);
	if (cc) return 0 < cc;
	return v1.X + v1.Y < v2.X + v2.Y;
}
PII subVec(PII v1, PII v2) {
	return { v1.X - v2.X, v1.Y - v2.Y };
}
bool _cmp(PII v1, PII v2) {
	int x1, x2, y1, y2;
	tie(x1, y1) = subVec(v1, pts[0]); 
	tie(x2, y2) = subVec(v2, pts[0]);
	if (1ll * y1 * x2 - 1ll * x1 * y2)
		return 1ll * y1 * x2 < 1ll * x1 * y2;
	else if (y1 - y2) return y1 < y2;
	return x1 < x2;
}
int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0);
	
	cin >> ptN;
	int a, b;
	FOR(i, 0, ptN) {
		cin >> a >> b;
		pts[i] = { a, b };
	}
	sort(pts, pts + ptN);
	sort(pts + 1, pts + ptN, cmp);
	stk[0] = pts[0];
	stk[1] = pts[1];
	stkIdx = 1;
	FOR(idx, 2, ptN) {
		PII pt = pts[idx];
		while (1 <= stkIdx && ccw(stk[stkIdx - 1], stk[stkIdx], pt) <= 0) {
			stkIdx--;
		}
		stk[++stkIdx] = pt;
	}
	cout << stkIdx + 1;
	//cout << '\n';
	//FOR(i, 0, stkIdx+1) cout << stk[i].X << ' ' << stk[i].Y << '\n';
}